File: android/app/src/main/kotlin/com/oreon/polygone_app/MdnsNativeImpl.kt
// MdnsNativeImpl.kt
// 
// Android native implementation using NSD (Network Service Discovery) API
// Handles service advertisement and discovery via mDNS/Bonjour
// 
// Location: android/app/src/main/kotlin/com/oreon/polygone_app/MdnsNativeImpl.kt

package com.oreon.polygone_app

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.net.wifi.WifiManager
import android.os.Build
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

/**
 * Android native implementation for mDNS service discovery
 * 
 * Responsibilities:
 * - Service advertisement via NsdManager.registerService()
 * - Service discovery via NsdManager.discoverServices()
 * - Multicast lock management for all-device discovery
 * - Callback routing to Flutter via MethodChannel
 */
class MdnsNativeImpl(private val context: Context, private val flutterEngine: FlutterEngine) {

    companion object {
        const val CHANNEL = "com.oreon.polygone_app/mdns"
        const val SERVICE_TYPE = "_oreonchat._tcp"
    }

    private val nsdManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    private val wifiManager = context.getApplicationContext()
        .getSystemService(Context.WIFI_SERVICE) as WifiManager
    
    private var multicastLock: WifiManager.MulticastLock? = null
    private var registrationListener: NsdManager.RegistrationListener? = null
    private var discoveryListener: NsdManager.DiscoveryListener? = null
    
    private var registeredServiceName: String? = null
    private val resolvedServices = mutableMapOf<String, ServiceInfo>()

    private data class ServiceInfo(
        val name: String,
        val address: String,
        val port: Int
    )

    /**
     * Sets up the method channel for Dart communication
     */
    fun setupChannel() {
        val channel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler { call, result ->
            try {
                when (call.method) {
                    "mdns.advertise" -> {
                        val serviceName = call.argument<String>("serviceName")
                        val serviceType = call.argument<String>("serviceType")
                        val port = call.argument<Int>("port")
                        
                        if (serviceName != null && serviceType != null && port != null) {
                            startAdvertisement(serviceName, serviceType, port)
                            result.success(null)
                        } else {
                            result.error("INVALID_ARGS", "Missing required arguments", null)
                        }
                    }
                    "mdns.discover" -> {
                        val serviceType = call.argument<String>("serviceType")
                        if (serviceType != null) {
                            startDiscovery(serviceType)
                            result.success(null)
                        } else {
                            result.error("INVALID_ARGS", "Missing serviceType", null)
                        }
                    }
                    "mdns.stop" -> {
                        stopMdns()
                        result.success(null)
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                result.error("ERROR", e.message, null)
            }
        }
    }

    /**
     * Starts service advertisement
     * 
     * Creates NsdServiceInfo and registers with NsdManager.
     * Acquires multicast lock for reliable discovery.
     */
    private fun startAdvertisement(serviceName: String, serviceType: String, port: Int) {
        // Create service info
        val serviceInfo = NsdServiceInfo().apply {
            this.serviceName = serviceName
            this.serviceType = serviceType
            this.port = port
        }

        // Acquire multicast lock
        acquireMulticastLock()

        // Create registration listener
        registrationListener = object : NsdManager.RegistrationListener {
            override fun onServiceRegistered(nsdServiceInfo: NsdServiceInfo) {
                registeredServiceName = nsdServiceInfo.serviceName
                sendCallbackToFlutter("onRegistered", mapOf(
                    "serviceName" to nsdServiceInfo.serviceName
                ))
            }

            override fun onRegistrationFailed(nsdServiceInfo: NsdServiceInfo, errorCode: Int) {
                sendCallbackToFlutter("onRegistrationFailed", mapOf(
                    "errorCode" to errorCode
                ))
            }

            override fun onServiceUnregistered(nsdServiceInfo: NsdServiceInfo) {
                registeredServiceName = null
                sendCallbackToFlutter("onUnregistered", emptyMap())
            }

            override fun onUnregistrationFailed(nsdServiceInfo: NsdServiceInfo, errorCode: Int) {
                sendCallbackToFlutter("onUnregistrationFailed", mapOf(
                    "errorCode" to errorCode
                ))
            }
        }

        // Register the service
        nsdManager.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, registrationListener)
    }

    /**
     * Starts service discovery
     * 
     * Creates DiscoveryListener and searches for services of the specified type.
     * When services are found, automatically resolves them to get IP addresses.
     */
    private fun startDiscovery(serviceType: String) {
        // Acquire multicast lock
        acquireMulticastLock()

        // Create discovery listener
        discoveryListener = object : NsdManager.DiscoveryListener {
            override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
                sendCallbackToFlutter("onDiscoveryStartFailed", mapOf(
                    "errorCode" to errorCode
                ))
            }

            override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
                sendCallbackToFlutter("onDiscoveryStopFailed", mapOf(
                    "errorCode" to errorCode
                ))
            }

            override fun onServiceFound(nsdServiceInfo: NsdServiceInfo) {
                // Resolve the discovered service to get address and port
                if (!nsdServiceInfo.serviceName.equals(registeredServiceName, ignoreCase = true)) {
                    nsdManager.resolveService(nsdServiceInfo, object : NsdManager.ResolveListener {
                        override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
                            sendCallbackToFlutter("onResolveFailed", mapOf(
                                "serviceName" to serviceInfo.serviceName,
                                "errorCode" to errorCode
                            ))
                        }

                        override fun onServiceResolved(nsdServiceInfo: NsdServiceInfo) {
                            val address = nsdServiceInfo.host?.hostAddress ?: return
                            val port = nsdServiceInfo.port
                            val serviceName = nsdServiceInfo.serviceName

                            resolvedServices[serviceName] = ServiceInfo(serviceName, address, port)

                            // Send callback to Flutter
                            sendCallbackToFlutter("onServiceDiscovered", mapOf(
                                "serviceName" to serviceName,
                                "address" to address,
                                "port" to port
                            ))
                        }
                    })
                }
            }

            override fun onServiceLost(nsdServiceInfo: NsdServiceInfo) {
                val serviceName = nsdServiceInfo.serviceName
                resolvedServices.remove(serviceName)

                sendCallbackToFlutter("onServiceLost", mapOf(
                    "serviceName" to serviceName
                ))
            }
        }

        // Start discovery
        nsdManager.discoverServices(serviceType, NsdManager.PROTOCOL_DNS_SD, discoveryListener)
    }

    /**
     * Stops all mDNS operations
     * 
     * Unregisters service, stops discovery, and releases multicast lock.
     */
    private fun stopMdns() {
        try {
            // Unregister service
            if (registrationListener != null) {
                nsdManager.unregisterService(registrationListener)
                registrationListener = null
            }

            // Stop discovery
            if (discoveryListener != null) {
                nsdManager.stopServiceDiscovery(discoveryListener)
                discoveryListener = null
            }

            // Clear resolved services
            resolvedServices.clear()

            // Release multicast lock
            releaseMulticastLock()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * Acquires multicast lock for mDNS operations
     * 
     * Multicast lock is required on Android 7+ for reliable mDNS discovery.
     * Without it, devices may not receive multicast packets.
     */
    private fun acquireMulticastLock() {
        if (multicastLock == null) {
            multicastLock = wifiManager.createMulticastLock("oreon_mdns")
            multicastLock?.acquire()
        }
    }

    /**
     * Releases the multicast lock
     */
    private fun releaseMulticastLock() {
        multicastLock?.release()
        multicastLock = null
    }

    /**
     * Sends callback to Flutter via method channel
     */
    private fun sendCallbackToFlutter(method: String, arguments: Map<String, Any?>) {
        try {
            val channel = MethodChannel(
                com.oreon.polygone_app.MainActivity.flutterEngineRef?.dartExecutor?.binaryMessenger,
                CHANNEL
            )
            channel.invokeMethod(method, arguments)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}

/**
 * Integration in MainActivity.kt:
 * 
 * Add this to your MainActivity class:
 * 
 * companion object {
 *     var flutterEngineRef: FlutterEngine? = null
 * }
 * 
 * override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
 *     super.configureFlutterEngine(flutterEngine)
 *     flutterEngineRef = flutterEngine
 *     
 *     val mdnsImpl = MdnsNativeImpl(this, flutterEngine)
 *     mdnsImpl.setupChannel()
 * }
 * 
 * Required permissions in AndroidManifest.xml:
 * - android.permission.INTERNET
 * - android.permission.ACCESS_WIFI_STATE
 * - android.permission.CHANGE_WIFI_MULTICAST_STATE
 */

File: ios/Runner/MdnsNativeImpl.swift
// MdnsNativeImpl.swift
//
// iOS native implementation using Bonjour (NSNetService/NSNetServiceBrowser)
// Handles service advertisement and discovery via mDNS
//
// Location: ios/Runner/MdnsNativeImpl.swift

import Foundation

/**
 * iOS mDNS implementation using Bonjour
 *
 * Responsibilities:
 * - Service advertisement via NSNetService
 * - Service discovery via NSNetServiceBrowser
 * - Callback routing to Flutter via MethodChannel
 */
class MdnsNativeImpl: NSObject {
    
    static let channelName = "com.oreon.polygone_app/mdns"
    static let serviceType = "_oreonchat._tcp"
    static let serviceDomain = "local."
    
    private var methodChannel: FlutterMethodChannel?
    private var netService: NetServiceWrapper?
    private var netServiceBrowser: NetServiceBrowserWrapper?
    private var discoveredServices: [String: ServiceInfo] = [:]
    
    struct ServiceInfo {
        let name: String
        let address: String
        let port: Int
    }
    
    /**
     * Initializes the mDNS implementation with Flutter method channel
     */
    func setupChannel(controller: FlutterViewController) {
        methodChannel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: controller.binaryMessenger
        )
        
        methodChannel?.setMethodCallHandler { [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) in
            switch call.method {
            case "mdns.advertise":
                if let args = call.arguments as? [String: Any],
                   let serviceName = args["serviceName"] as? String,
                   let port = args["port"] as? Int {
                    self?.startAdvertisement(serviceName: serviceName, port: port)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                }
                
            case "mdns.discover":
                self?.startDiscovery()
                result(nil)
                
            case "mdns.stop":
                self?.stopMdns()
                result(nil)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    /**
     * Starts service advertisement
     *
     * Creates NSNetService and publishes it for discovery by other devices
     */
    private func startAdvertisement(serviceName: String, port: Int) {
        netService = NetServiceWrapper(
            name: serviceName,
            type: Self.serviceType,
            domain: Self.serviceDomain,
            port: Int32(port),
            methodChannel: methodChannel
        )
        netService?.publish()
    }
    
    /**
     * Starts service discovery
     *
     * Creates NSNetServiceBrowser and searches for services of the specified type
     */
    private func startDiscovery() {
        netServiceBrowser = NetServiceBrowserWrapper(
            serviceType: Self.serviceType,
            domain: Self.serviceDomain,
            methodChannel: methodChannel,
            onServiceDiscovered: { [weak self] serviceName, address, port in
                self?.discoveredServices[serviceName] = ServiceInfo(
                    name: serviceName,
                    address: address,
                    port: port
                )
            },
            onServiceLost: { [weak self] serviceName in
                self?.discoveredServices.removeValue(forKey: serviceName)
            }
        )
        netServiceBrowser?.browse()
    }
    
    /**
     * Stops all mDNS operations
     */
    private func stopMdns() {
        netService?.stop()
        netService = nil
        
        netServiceBrowser?.stop()
        netServiceBrowser = nil
        
        discoveredServices.removeAll()
    }
}

/**
 * Wrapper for NSNetService to handle service advertisement
 */
class NetServiceWrapper: NSObject, NetServiceDelegate {
    
    private let netService: NetService
    private let methodChannel: FlutterMethodChannel?
    
    init(name: String, type: String, domain: String, port: Int32, methodChannel: FlutterMethodChannel?) {
        self.netService = NetService(domain: domain, type: type, name: name, port: port)
        self.methodChannel = methodChannel
        super.init()
        self.netService.delegate = self
    }
    
    func publish() {
        netService.publish(options: .listenForConnections)
    }
    
    func stop() {
        netService.stop()
    }
    
    // MARK: - NetServiceDelegate
    
    func netServiceDidPublish(_ sender: NetService) {
        methodChannel?.invokeMethod("onRegistered", arguments: [
            "serviceName": sender.name
        ])
    }
    
    func netService(_ sender: NetService, didNotPublish errorDict: [String: NSNumber]) {
        if let errorCode = errorDict[NetService.errorCode]?.intValue {
            methodChannel?.invokeMethod("onRegistrationFailed", arguments: [
                "errorCode": errorCode
            ])
        }
    }
    
    func netServiceDidStop(_ sender: NetService) {
        methodChannel?.invokeMethod("onUnregistered", arguments: [:])
    }
}

/**
 * Wrapper for NSNetServiceBrowser to handle service discovery
 */
class NetServiceBrowserWrapper: NSObject, NetServiceBrowserDelegate, NetServiceDelegate {
    
    private let browser: NetServiceBrowser
    private let serviceType: String
    private let domain: String
    private let methodChannel: FlutterMethodChannel?
    private let onServiceDiscovered: (String, String, Int) -> Void
    private let onServiceLost: (String) -> Void
    private var resolvedServices: [String: Net ServiceInfo] = [:]
    
    struct NetServiceInfo {
        let service: NetService
        let name: String
    }
    
    init(serviceType: String,
         domain: String,
         methodChannel: FlutterMethodChannel?,
         onServiceDiscovered: @escaping (String, String, Int) -> Void,
         onServiceLost: @escaping (String) -> Void) {
        
        self.browser = NetServiceBrowser()
        self.serviceType = serviceType
        self.domain = domain
        self.methodChannel = methodChannel
        self.onServiceDiscovered = onServiceDiscovered
        self.onServiceLost = onServiceLost
        
        super.init()
        self.browser.delegate = self
    }
    
    func browse() {
        browser.searchForServices(ofType: serviceType, inDomain: domain)
    }
    
    func stop() {
        browser.stop()
        resolvedServices.removeAll()
    }
    
    // MARK: - NetServiceBrowserDelegate
    
    func netServiceBrowserWillSearch(_ browser: NetServiceBrowser) {
        // Discovery started
    }
    
    func netServiceBrowserDidStopSearch(_ browser: NetServiceBrowser) {
        // Discovery stopped
    }
    
    func netServiceBrowser(_ browser: NetServiceBrowser,
                         didFindDomain domainString: String,
                         moreComing: Bool) {
        // Domain found (not used for our simple use case)
    }
    
    func netServiceBrowser(_ browser: NetServiceBrowser,
                         didFind netService: NetService,
                         moreComing: Bool) {
        // Service found - resolve it to get address and port
        let serviceName = netService.name
        
        // Skip duplicate resolution attempts
        if resolvedServices[serviceName] != nil {
            return
        }
        
        // Create resolver wrapper
        let resolver = netService
        resolver.delegate = self
        resolver.resolve(withTimeout: 5.0)
        
        // Track this service
        resolvedServices[serviceName] = NetServiceInfo(service: resolver, name: serviceName)
    }
    
    func netServiceBrowser(_ browser: NetServiceBrowser,
                         didRemove netService: NetService,
                         moreComing: Bool) {
        let serviceName = netService.name
        resolvedServices.removeValue(forKey: serviceName)
        
        methodChannel?.invokeMethod("onServiceLost", arguments: [
            "serviceName": serviceName
        ])
        
        onServiceLost(serviceName)
    }
    
    func netServiceBrowserDidStopSearch(_ browser: NetServiceBrowser,
                                       error errorDict: [String: NSNumber]) {
        if let errorCode = errorDict[NetService.errorCode]?.intValue {
            methodChannel?.invokeMethod("onDiscoveryStopFailed", arguments: [
                "errorCode": errorCode
            ])
        }
    }
    
    // MARK: - NetServiceDelegate (for resolution)
    
    func netServiceDidResolveAddress(_ sender: NetService) {
        guard let addresses = sender.addresses else { return }
        
        // Extract IPv4 address from addresses
        var ipAddress: String?
        for address in addresses {
            ipAddress = addressToString(address)
            if ipAddress != nil && !ipAddress!.contains(":") {
                // Prefer IPv4
                break
            }
        }
        
        guard let address = ipAddress else { return }
        
        let serviceName = sender.name
        let port = sender.port
        
        methodChannel?.invokeMethod("onServiceDiscovered", arguments: [
            "serviceName": serviceName,
            "address": address,
            "port": port
        ])
        
        onServiceDiscovered(serviceName, address, Int(port))
    }
    
    func netService(_ sender: NetService, didNotResolve errorDict: [String: NSNumber]) {
        if let errorCode = errorDict[NetService.errorCode]?.intValue {
            methodChannel?.invokeMethod("onResolveFailed", arguments: [
                "serviceName": sender.name,
                "errorCode": errorCode
            ])
        }
    }
    
    /**
     * Converts NSData address to string representation
     */
    private func addressToString(_ address: Data) -> String? {
        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
        
        guard address.withUnsafeBytes({ bytes -> Int32 in
            getnameinfo(
                bytes.baseAddress?.assumingMemoryBound(to: sockaddr.self),
                socklen_t(address.count),
                &hostname,
                socklen_t(hostname.count),
                nil,
                0,
                NI_NUMERICHOST
            )
        }) == 0 else {
            return nil
        }
        
        return String(cString: hostname)
    }
}

/**
 * Integration in GeneratedPluginRegistrant or main.swift:
 *
 * In your GeneratedPluginRegistrant.swift or main view controller:
 *
 * override func application(
 *     _ application: UIApplication,
 *     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
 * ) -> Bool {
 *     if let controller = window?.rootViewController as? FlutterViewController {
 *         let mdnsImpl = MdnsNativeImpl()
 *         mdnsImpl.setupChannel(controller: controller)
 *     }
 *     return true
 * }
 *
 * Required Info.plist entries:
 * - NSBonjourServices: [_oreonchat._tcp, _oreonchat._udp]
 * - NSLocalNetworkUsageDescription: "Oreon needs access to your local network
 *   to discover and communicate with nearby devices for LAN chat."
 * - NSBonjourUsageDescription: "Oreon uses Bonjour to discover nearby devices."
 *
 * Optional entries:
 * - UIBackgroundModes: [fetch, processing] (for background operation)
 * - NSAllowsLocalNetworking: true (under NSAppTransportSecurity)
 */
